src/wallet/scriptpubkeyman.cpp

```diff
@@ -1414,30 +1415,44 @@ bool LegacyScriptPubKeyMan::TopUpInner(unsigned int kpSize)
         {
             // don't create extra internal keys
             missingInternal = 0;
-        } else {
-            nTargetSize *= 2;
         }
+        int64_t total_missing = missingInternal + missingExternal;
+        if (total_missing == 0) return true;
+        bool should_show_progress = total_missing > 100;
         bool fInternal = false;
         WalletBatch batch(m_storage.GetDatabase());
-        for (int64_t i = missingInternal + missingExternal; i--;)
+        std::string strMsg = _("Topping up keypool...").translated;
+        WalletLogPrintf("%s\n", strMsg);
+        if (should_show_progress) {
+            uiInterface.ShowProgress(strMsg, 0, false);
+        }
+        int64_t progress_report_time = GetTime();
+        constexpr int64_t PROGRESS_REPORT_INTERVAL = 1; // in seconds
+        int64_t i{0};
```

Comment:
nit: so far as `i` is used outside of loop maybe give it a better name? Otherwise it feels like LogPrintf is still inside a loop.